### PR TITLE
=act split Props java api implementation in separate files.

### PR DESCRIPTION
**Pull request by andreaTP**
This require a split into 3 compilation units, but no semantic modifications.

### CODE REVIEW

**Comment by hepin1989** [comment]
how about make ```producer``` a private[akka] api?

**Comment by patriknw** [comment]
> how about make producer a private[akka] api?

What do you mean? IndirectActorProducer?

**Comment by patriknw** [comment]
OK TO TEST

**Comment by hepin1989** [comment]
yes

**Comment by andreaTP** [comment]
@hepin1989  I just ported the existing version, I think that the topic goes outside this PR purpose.

**Comment by ktoso** [comment]
Yep, let's keep this PR a simple file move


-- 
Cheers,
Konrad 'ktoso’ Malawski
Akka @ Typesafe

On 11 August 2015 at 16:48:39, Andrea (notifications@github.com) wrote:

@hepin1989 I just ported the existing version, I think that the topic goes outside this PR purpose.

**Comment by patriknw** [comment]
@hepin1989 IndirectActorProducer is important for dependency injection, http://doc.akka.io/docs/akka/2.3.12/java/untyped-actors.html#Dependency_Injection

**Comment by andreaTP** [comment]
ok, able to reproduce, is a problem with Java interop.
Will fix it, have I got to rebase at the end?

**Comment by andreaTP** [comment]
fixed and rebased, ready to re-test.

Sorry.

File: akka-actor/src/main/scala/akka/actor/Props.scala [hunk 1]
```diff
@@ -156,10 +112,9 @@ object Props {
  * }}}
  */
 @SerialVersionUID(2L)
-final case class Props(deploy: Deploy, clazz: Class[_], args: immutable.Seq[Any]) {
+final case class Props(deploy: Deploy = Props.defaultDeploy, clazz: Class[_], args: immutable.Seq[Any]) {
 
-  if (Modifier.isAbstract(clazz.getModifiers))
-    throw new IllegalArgumentException(s"Actor class [${clazz.getName}] must not be abstract")
+  Props.isAbstract(clazz)
```
Comment (ktoso) [inline akka-actor/src/main/scala/akka/actor/Props.scala]: :-1: for making an `isSomething` method throw;

Seems this PR is not only "move" but some new methods?

File: akka-actor/src/main/scala/akka/actor/AbstractProps.scala [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+import akka.japi.Util.immutableSeq
+
+/**
+ * Java API: Factory for Props instances.
+ */
+trait AbstractProps {
+
+  def isAbstract(clazz: Class[_]) =
+    if (Modifier.isAbstract(clazz.getModifiers))
+      throw new IllegalArgumentException(s"Actor class [${clazz.getName}] must not be abstract")
```
Comment (ktoso) [inline akka-actor/src/main/scala/akka/actor/AbstractProps.scala]: "isSomething" feels POJO getter-like, would avoid these method names for validation methods

File: akka-actor/src/main/scala/akka/actor/Props.scala (hunk 1 above)
Comment (andreaTP) [inline akka-actor/src/main/scala/akka/actor/Props.scala]: ```validate``` is ok to me!

I had to make some really tiny modifications to keep everything working properly, this is not strictly needed, but this is spare code in the body of a class, refactoring it out to a companion object method could be considered good practice, moreover it is in the AbstractProps because is a really JVM dependant check.
Does this sounds good to you?

**Comment by andreaTP** [comment]
Do not know your PRs rules, have I got to squash commits into one or is ok to leave the history of the changes?

**Comment by ktoso** [comment]
The contributing guidelines are written up here: https://github.com/akka/akka/blob/master/CONTRIBUTING.md

Yes, please squash and make the commit name match our style there (`=act split things into more files` (example name))

File: akka-actor/src/main/scala/akka/actor/Props.scala [hunk 3]
```diff
@@ -156,10 +112,9 @@ object Props {
  * }}}
  */
 @SerialVersionUID(2L)
-final case class Props(deploy: Deploy, clazz: Class[_], args: immutable.Seq[Any]) {
+final case class Props(deploy: Deploy = Props.defaultDeploy, clazz: Class[_], args: immutable.Seq[Any]) {
```
Comment (ktoso) [inline akka-actor/src/main/scala/akka/actor/Props.scala]: Why is this needed?
In general adding default params is better in the latter (not initial) parameters

File: akka-actor/src/main/scala/akka/actor/Props.scala [hunk 4]
```diff
@@ -156,10 +112,9 @@ object Props {
...
 
-  if (Modifier.isAbstract(clazz.getModifiers))
-    throw new IllegalArgumentException(s"Actor class [${clazz.getName}] must not be abstract")
+  Props.validate(clazz)
```
Comment (ktoso) [inline akka-actor/src/main/scala/akka/actor/Props.scala]: ok, that's better I guess

File: akka-actor/src/main/scala/akka/actor/Props.scala (hunk 3 above)
Comment (andreaTP) [inline akka-actor/src/main/scala/akka/actor/Props.scala]: Actually this is no more needed, so I remove this and put the modification into ```AbstractProps```.
will see this into the next commit.

**Comment by andreaTP** [comment]
ok, hope this pr looks better now

File: akka-actor/src/main/scala/akka/actor/Props.scala (hunk 3 above)
Comment (ktoso) [inline akka-actor/src/main/scala/akka/actor/Props.scala]: great :)

**Comment by ktoso** [comment]
LGTM if validation is happy :)

**Comment by andreaTP** [comment]
PLS BUILD

**Comment by andreaTP** [comment]
Am I supposed to do anything else to get this merged?

**Comment by ktoso** [comment]
I'd like another pair of eyes on it but from my perspective LGTM,
we'll merge soon (currently rushing to release 2.4-M3 and focusing on persistence tickets, sorry for the dalay)

**Comment by andreaTP** [comment]
perfectly ok with this, thanks a lot! 
The question was only because these are my first commits here and I'm not sure on the workflow here :-)

**Comment by ktoso** [comment]
I see, the workflow usually is that we merge after a PR gets "enough" LGTMs, where "enough" may vary on complexity of the PR. This one seems rather simple, yet still there was some small nuances that are now fixed. I'd like a second pair of eyes on it before we merge :)

File: akka-actor/src/main/scala/akka/actor/AbstractProps.scala [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+import scala.language.existentials
+import scala.reflect.ClassTag
+import akka.japi.Util.immutableSeq
+
+/**
+ * Java API: Factory for Props instances.
+ */
+trait AbstractProps {
+
+  def validate(clazz: Class[_]) =
```
Comment (patriknw) [inline akka-actor/src/main/scala/akka/actor/AbstractProps.scala]: why public

File: akka-actor/src/main/scala/akka/actor/AbstractProps.scala [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+
+import scala.annotation.varargs
+import scala.language.existentials
+import scala.reflect.ClassTag
+import akka.japi.Util.immutableSeq
+
+/**
+ * Java API: Factory for Props instances.
+ */
+trait AbstractProps {
```
Comment (patriknw) [inline akka-actor/src/main/scala/akka/actor/AbstractProps.scala]: what is the motivation for splitting it?

File: akka-actor/src/main/scala/akka/actor/AbstractProps.scala (hunk 5 above)
Comment (andreaTP) [inline akka-actor/src/main/scala/akka/actor/AbstractProps.scala]: just because Props is public too, if private works for you I can go for it

Comment (ktoso) [inline akka-actor/src/main/scala/akka/actor/AbstractProps.scala]: let's private it :)

File: akka-actor/src/main/scala/akka/actor/AbstractProps.scala (hunk 6 above)
Comment (ktoso) [inline akka-actor/src/main/scala/akka/actor/AbstractProps.scala]: https://github.com/unicredit/akka.js :wink:

File: akka-actor/src/main/scala/akka/actor/AbstractProps.scala (hunk 5 above)
Comment (andreaTP) [inline akka-actor/src/main/scala/akka/actor/AbstractProps.scala]: ok

File: akka-actor/src/main/scala/akka/actor/AbstractProps.scala (hunk 6 above)
Comment (andreaTP) [inline akka-actor/src/main/scala/akka/actor/AbstractProps.scala]: I also believe that code separation jvm dependant / agnostic framework is a good design just in style with this great project.
More projects can benefits from this in the future.

**Comment by andreaTP** [comment]
```private```'d and squashed commits

**Comment by ktoso** [comment]
Refs https://github.com/akka/akka/issues/18234

**Comment by ktoso** [comment]
Superseeded by https://github.com/akka/akka/pull/18238 (thanks!)
